admission/hooks/validate_deployment_replica_count_test.go

```diff
@@ -0,0 +1,146 @@
+package hooks
+
+import (
+	"strings"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+var _ = Describe("validate deployment replica count webhook", func() {
+	It("should allow a Deployment with replicas = 0", func() {
+		deployManifest := `apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: ubuntu-deployment-1
+  namespace: default
+  labels:
+    app: ubuntu
+  annotations:
+    admission.cybozu.com/force-replica-count: "0"
+spec:
+  replicas: 0
+  selector:
+    matchLabels:
+      app: ubuntu
+  template:
+    metadata:
+      labels:
+        app: ubuntu
+    spec:
+      containers:
+      - name: ubuntu
+        image: quay.io/cybozu/ubuntu
+`
+
+		d := yaml.NewYAMLOrJSONDecoder(strings.NewReader(deployManifest), 4096)
+		deploy := &appsv1.Deployment{}
+		err := d.Decode(deploy)
+		Expect(err).NotTo(HaveOccurred())
+
+		err = k8sClient.Create(testCtx, deploy)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("should deny a Deployment with replicas != 0", func() {
+		deployManifest := `apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: ubuntu-deployment-2
+  namespace: default
+  labels:
+    app: ubuntu
+  annotations:
+    admission.cybozu.com/force-replica-count: "0"
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: ubuntu
+  template:
+    metadata:
+      labels:
+        app: ubuntu
+    spec:
+      containers:
+      - name: ubuntu
+        image: quay.io/cybozu/ubuntu
+`
+
+		d := yaml.NewYAMLOrJSONDecoder(strings.NewReader(deployManifest), 4096)
+		deploy := &appsv1.Deployment{}
+		err := d.Decode(deploy)
+		Expect(err).NotTo(HaveOccurred())
+
+		err = k8sClient.Create(testCtx, deploy)
+		Expect(err).To(HaveOccurred())
+	})
+
+	It("should allow a Deployment without force-replica-count annotation", func() {
+		deployManifest := `apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: ubuntu-deployment-3
+  namespace: default
+  labels:
+    app: ubuntu
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: ubuntu
+  template:
+    metadata:
+      labels:
+        app: ubuntu
+    spec:
+      containers:
+      - name: ubuntu
+        image: quay.io/cybozu/ubuntu
+`
+
+		d := yaml.NewYAMLOrJSONDecoder(strings.NewReader(deployManifest), 4096)
+		deploy := &appsv1.Deployment{}
+		err := d.Decode(deploy)
+		Expect(err).NotTo(HaveOccurred())
+
+		err = k8sClient.Create(testCtx, deploy)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("should allow a Deployment with annotation force-replica-count != 0", func() {
+		deployManifest := `apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: ubuntu-deployment-4
+  namespace: default
+  labels:
+    app: ubuntu
+  annotations:
+    admission.cybozu.com/force-replica-count: "2"
+spec:
+  replicas: 1
```

Comment:
It would be good to clarify that this Deployment is erroneous even though force-replica-count == .spec.replicas.

```suggestion
    admission.cybozu.com/force-replica-count: "1"
spec:
  replicas: 1
```

Reply:
Changed to test both cases of `force-replica-count == .spec.replicas.` and `force-replica-count != .spec.replicas.`.